codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoderForBuffer.java

```diff
@@ -399,17 +399,20 @@ protected void decodeLast(ChannelHandlerContext ctx, Buffer in) throws Exception
         }
     }
 
-    private static Buffer expandCumulation(BufferAllocator alloc, Buffer oldCumulation, Buffer in) {
-        Buffer newCumulation =
-                alloc.allocate(safeFindNextPositivePowerOfTwo(oldCumulation.readableBytes() + in.readableBytes()));
-        Buffer toRelease = newCumulation;
+    private static Buffer expandCumulationAndWrite(BufferAllocator alloc, Buffer oldCumulation, Buffer in) {
+        final int newSize = safeFindNextPositivePowerOfTwo(oldCumulation.readableBytes() + in.readableBytes());
+        Buffer newCumulation = oldCumulation.readOnly() ? alloc.allocate(newSize) :
+                oldCumulation.ensureWritable(newSize, in.readableBytes(), true);
```

Comment:
The second argument to `ensureWritable` is the number of bytes that the buffer should grow its capacity by, if it needs to grow. It would need to grow if there is not already room, and compaction cannot make enough room, but since the third argument is `true`, compaction would still happen. So that means we'll get some room from compaction, and some more from growing the buffer. If we use `oldCumulation.capacity()` here, then the cumulation buffer capacity will _at least_ double, if and when it grows. The growth will always ensure that `newSize` will fit in the expanded buffer.

My point is that growing by at least `in.readableBytes()` is a bit of an odd choice, that will grow the buffer not following any pattern, but rather as a consequence of the sequence of reads, and the sizes of the readable chunks of the input buffers. These are two unpredictable things that will result in the cumulation buffer having an unpredictable size.

Reply:
Let me just use the default `ensureWritable` for now.